integrations/ad_integration/ad_sync.py

```diff
@@ -346,24 +365,108 @@ def _edit_user_addresses(self, uuid, ad_object):
+            response = self.helper._mo_post('details/terminate', payload)
+            logger.debug('Response: {}'.format(response.text))
+
+    def _finalize_user_addresses(self, uuid, ad_object):
+        if 'user_addresses' not in self.mapping:
+            return
+
+        today = datetime.strftime(datetime.now(), "%Y-%m-%d")
+        fields_to_edit = self._find_existing_ad_address_types(uuid)
+
+        def check_ad_field_exists(field):
+            if field not in ad_object:
+                logger.debug('No such AD field: {}'.format(field))
+                return False
+            return True
+
+        def check_field_in_fields_to_edit(field):
+            return field in fields_to_edit.keys()
+
+        def check_validity_is_ok(field):
+            # NOTE: Maybe this should be not set, or in the future?
+            return fields_to_edit[field]['validity']['to'] is None
+
+        # Find fields to terminate
+        address_fields = self.mapping['user_addresses'].keys()
+        address_fields = filter(check_ad_field_exists, address_fields)
+        address_fields = filter(check_field_in_fields_to_edit, address_fields)
+        address_fields = filter(check_validity_is_ok, address_fields)
+        for field in address_fields:
+            payload = {
+                'type': 'address',
+                'uuid': fields_to_edit[field]['uuid'],
+                'validity': {"to": today}
+            }
+            logger.debug('Finalize payload: {}'.format(payload))
+            response = self.helper._mo_post('details/terminate', payload)
+            logger.debug('Response: {}'.format(response.text))
+
     def _update_single_user(self, uuid, ad_object):
         """
         Update all fields for a single user.
         :param uuid: uuid of the user.
         :param ad_object: Dict with the AD information for the user.
         """
+        # Debug log if enabled is not found
+        if 'Enabled' not in ad_object:
+            logger.info('{} not in ad_object'.format("Enabled"))
+
+        # Lookup whether or not to syncronize disabled users
+        sync_disabled = self.settings.get("integrations.ad.ad_mo_sync_disabled", True)
+        # Check whether the current user is disabled or not
+        current_user_is_disabled = ad_object.get('Enabled', True) == False
+        if sync_disabled == False and current_user_is_disabled:
+            finalize_key = 'integrations.ad.ad_mo_sync_finalize_disabled' 
+            if finalize_key not in self.settings:
+                raise Exception("{} not in settings".format(finalize_key))
+            # If configured to finalize
+            if self.settings[finalize_key]:
+                # Set validity end --> today if in the future
+                self._finalize_it_system(uuid, ad_object)
+                self._finalize_user_addresses(uuid, ad_object)
+            # Otherwise do not do anything
+            return
+
+        # Sync the user, whether disabled or not
```

Comment:
Aren't we supposed to not import these details if the user is disabled?

Reply:
There is an early return on 467